3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,248 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import './polyfills';
+import {dev} from '../src/log';
+import {initLogConstructor, setReportError} from '../src/log';
+import {reportError} from '../src/error';
+import {IframeMessagingClient} from './iframe-messaging-client';
+import {MessageTypes} from '../src/3p-analytics-common';
+initLogConstructor();
+setReportError(reportError);
+
+/** @private @const {string} */
+const TAG_ = 'ampanalytics-lib';
+
+/**
+ * Receives messages bound for this cross-domain iframe, from all creatives
+ */
+class AmpAnalytics3pMessageRouter {
+  /** @param {!Window} win */
+  constructor(win) {
+
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {string} */
+    this.sentinel_ = null;
+
+    /** @private {Object<string, AmpAnalytics3pCreativeMessageRouter>} */
+    this.creativeMessageRouters_ = [];
+
+    /**
+     * Simple requestIdleCallback polyfill
+     * @type {!function(!function(), number)}
+     */
+    this.requestIdleCallback =
+      this.win_.requestIdleCallback.bind(this.win_) ||
+      (cb => setTimeout(cb, 1));
+
+    /**
+     * Handles communication between frames
+     * @private {!IframeMessagingClient}
+     */
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    try {
+      const frameNameData = JSON.parse(this.win_.name);
+      if (frameNameData.sentinel) {
+        this.sentinel_ = frameNameData.sentinel;
+        this.iframeMessagingClient_.setSentinel(this.sentinel_);
+      }
+    } catch (e) {
+      dev().warn(TAG_, 'Unable to set sentinel');
+      return;
+    }
+    this.iframeMessagingClient_.registerCallback(
+      MessageTypes.ampAnalytics3pMessages,
+      messages => {
+        // All the messages are for this frame, but may come from different
+        // creatives.
+        messages[MessageTypes.ampAnalytics3pMessages].forEach(msg => {
+          if (!this.creativeMessageRouters_[msg.senderId]) {
+            this.creativeMessageRouters_[msg.senderId] =
+              new AmpAnalytics3pCreativeMessageRouter(this, this.sentinel_,
+                msg.senderId);
+          }
+          this.creativeMessageRouters_[msg.senderId].receiveMessage(msg);
```

Comment:
It's an id used to identify the creative, yes. It's chosen in the Transport constructor.